src/snapred/backend/profiling/ProgressRecorder.py

```diff
@@ -0,0 +1,989 @@
+            module = inspect.getmodule(frame.f_code)
+            module_name = module.__name__ if module is not None else ""
+
+            qualified_name = None
+            if sys.version_info >= (3, 11, 0):
+                qualified_name = frame.f_code.co_qualname
+            else:
+                # Required if Python version < 3.11.
+
+                # Get the function name
+                func_name = frame.f_code.co_name
+
+                # If the function is a method within a class, try to find the class name
+                class_name = None
+                if "self" in frame.f_locals:
+                    class_name = frame.f_locals["self"].__class__.__qualname__
+                elif "cls" in frame.f_locals:
+                    class_name = frame.f_locals["cls"].__qualname__
+
+                qualified_name = class_name + "." + func_name
+
+        return (module_name, qualified_name)
+
+    @classmethod
+    def isLoggingEnabledForStep(cls, key: Tuple[str, ...]) -> bool:
+        # Determine whether or not to log a step, based on the qualified name of its scope.
+
+        # key components: (<module name>, <fully-qualified scope name>, <explicit step name>)
+        qualname = key[-2]
+        if qualname is not None:
+            _root = qualname.split(".")[0]
+            return _root in cls._loggable_qualname_roots()
+        return False
+
+    @classmethod
+    def getStepKey(
+        cls, *, callerOrStackFrameOverride: FunctionType | FrameType | type = None, stepName=None
+    ) -> Tuple[str | None, ...]:
+        # Compute a unique key for the current process step.
+
+        # Notes:
+        #   * In general, the key tuple will be
+        #     `(<module name>, <fully qualified class or method name>, <explicit step name>)`.
+        #   * `None` is a valid entry for any tuple component;
+        #   * Note that the caller's stack frame is actually two-frames up from the current frame.
+
+        if bool(callerOrStackFrameOverride) and not isinstance(
+            callerOrStackFrameOverride, (FunctionType, FrameType, type)
+        ):
+            raise RuntimeError(
+                "Usage error: when `callerOrStackFrameOverride` is set, it must be either "
+                + "a function, the local stack-frame of a function, or a <class> type."
+            )
+
+        # The choice of stack frame in the fallback here allows `ProgressRecorder.record` to be called directly
+        # in the scope of a function of interest.  Any other usage should supply the <caller function>
+        # or <local stack-frame of caller function> directly.
+        # TODO: maybe there shouldn't be any fallback?
+        callerObjectOrStackFrame = (
+            callerOrStackFrameOverride
+            if callerOrStackFrameOverride is not None
+            else inspect.currentframe().f_back.f_back
+        )
+        key = cls._getCallerFullyQualifiedName(callerObjectOrStackFrame) + (stepName,)
```

Comment:
Could the key use the hash of the caller's code instead? It seems like we must go through some lengths to cover the qualname for each of these types.